File: backend/schema_models/watch.py
"""Pydantic models for watch folder monitoring operations."""

from __future__ import annotations

from pydantic import BaseModel, Field

from .convert import ConvertOptions


class WatchOptions(BaseModel):
    poll_interval: float = 2.0
    extensions: list[str] | None = None
    once: bool = False
    convert_opts: ConvertOptions = Field(default_factory=ConvertOptions)


class WatchStartRequest(BaseModel):
    session_id: str
    options: WatchOptions = Field(default_factory=WatchOptions)


class WatchStartResponse(BaseModel):
    watch_id: str
    source: str
    output: str


class WatchStopRequest(BaseModel):
    session_id: str


class WatchStopResponse(BaseModel):
    stopped: bool


class WatchStatusResponse(BaseModel):
    running: bool
    started_at: float | None = None
    source: str | None = None
    output: str | None = None
    files_processed: int = 0
    source_tokens: int = 0
    target_tokens: int = 0

File: frontend/lib/api/types/watch.ts
/**
 * Watcher streaming and WebSocket event API types.
 */

export interface WatchOptions {
  poll_interval?: number;
  extensions?: string[];
  once?: boolean;
  convert_opts?: Record<string, unknown>;
}

export interface WatchStartRequest {
  session_id: string;
  options?: WatchOptions;
}

export interface WatchStartResponse {
  watch_id: string;
  source: string;
  output: string;
}

export interface WatchStopResponse {
  stopped: boolean;
}

export interface WatchStatus {
  running: boolean;
  started_at?: string;
  source?: string;
  output?: string;
  files_processed?: number;
  source_tokens?: number;
  target_tokens?: number;
  percent?: number;
}

export interface WsEnvelope<T = unknown> {
  type: string;
  session_id?: string;
  ts?: number;
  data: T;
}

export interface WatchStartedData {
  source?: string;
  output?: string;
}

export interface WatchFileData {
  file: string;
  status: "queued" | "converting" | "done" | "skipped" | "error";
  source_tokens?: number;
  target_tokens?: number;
  percent?: number;
  error?: string;
}

export interface WatchTotalData {
  files: number;
  source_tokens: number;
  target_tokens: number;
  percent: number;
}

export interface WatchStoppedData {
  reason?: string;
}

export interface ProgressData {
  job_id: string;
  operation: string;
  current: number;
  total: number;
  file?: string;
  percent?: number;
}

export interface JobDoneData {
  job_id: string;
  operation: string;
  summary: unknown;
}

export interface LogData {
  level: string;
  message: string;
}

export type WatchEvent =
  | { type: "watch.started"; data: WatchStartedData }
  | { type: "watch.file"; data: WatchFileData }
  | { type: "watch.total"; data: WatchTotalData }
  | { type: "watch.stopped"; data: WatchStoppedData }
  | { type: "progress"; data: ProgressData }
  | { type: "job.done"; data: JobDoneData }
  | { type: "log"; data: LogData };
